staging/src/k8s.io/kube-apiextensions-server/pkg/controller/finalizer/crd_finalizer.go

```diff
@@ -0,0 +1,329 @@
+	// and the endpoint is serviceable.  if they aren't, then there's nothing to cleanup.
+	if !apiextensions.IsCRDConditionFalse(crd, apiextensions.NameConflict) {
+		apiextensions.SetCRDCondition(crd, apiextensions.CustomResourceDefinitionCondition{
+			Type:    apiextensions.Terminating,
+			Status:  apiextensions.ConditionTrue,
+			Reason:  "InstanceDeletionStuck",
+			Message: fmt.Sprintf("cannot proceed with deletion because of %v condition", apiextensions.NameConflict),
+		})
+		crd, err = c.crdClient.CustomResourceDefinitions().UpdateStatus(crd)
+		if err != nil {
+			return err
+		}
+		return fmt.Errorf("cannot proceed with deletion because of %v condition", apiextensions.NameConflict)
+	}
+
+	// Now we can start deleting items.  We should use the REST API to ensure that all normal admission runs.
+	// Since we control the endpoints, we know that delete collection works.
+	crClient, err := c.clientPool.ClientForGroupVersionResource(schema.GroupVersionResource{Group: crd.Spec.Group, Version: crd.Spec.Version, Resource: crd.Status.AcceptedNames.Plural})
+	if err != nil {
+		return err
+	}
+	crAPIResource := &metav1.APIResource{
+		Name:         crd.Status.AcceptedNames.Plural,
+		SingularName: crd.Status.AcceptedNames.Singular,
+		Namespaced:   crd.Spec.Scope == apiextensions.NamespaceScoped,
+		Kind:         crd.Status.AcceptedNames.Kind,
+		Verbs:        metav1.Verbs([]string{"deletecollection", "list"}),
+		ShortNames:   crd.Status.AcceptedNames.ShortNames,
+	}
+	crResourceClient := crClient.Resource(crAPIResource, "" /* namespace all */)
+	allResources, err := crResourceClient.List(metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	deletedNamespaces := sets.String{}
+	deleteErrors := []error{}
+	for _, item := range allResources.(*unstructured.UnstructuredList).Items {
+		metadata, err := meta.Accessor(&item)
+		if err != nil {
+			utilruntime.HandleError(err)
+			continue
+		}
+		if deletedNamespaces.Has(metadata.GetNamespace()) {
+			continue
+		}
+		// don't retry deleting the same namespace
+		deletedNamespaces.Insert(metadata.GetNamespace())
+		if err := crClient.Resource(crAPIResource, metadata.GetNamespace()).DeleteCollection(nil, metav1.ListOptions{}); err != nil {
+			deleteErrors = append(deleteErrors, err)
+			continue
+		}
+	}
+	if deleteError := utilerrors.NewAggregate(deleteErrors); deleteError != nil {
+		apiextensions.SetCRDCondition(crd, apiextensions.CustomResourceDefinitionCondition{
+			Type:    apiextensions.Terminating,
+			Status:  apiextensions.ConditionTrue,
+			Reason:  "InstanceDeletionFailed",
+			Message: fmt.Sprintf("could not issue all deletes: %v", deleteError),
+		})
+		crd, err = c.crdClient.CustomResourceDefinitions().UpdateStatus(crd)
+		if err != nil {
+			utilruntime.HandleError(err)
+		}
+		return deleteError
+	}
+
+	// now we need to wait until all the resources are deleted.  Start with a simple poll before we do anything fancy
```

Comment:
> object gets created because the cache used by the CR handler has not yet observed the Terminating condition update on the CRD. Perhaps it's not worth addressing now, but maybe TODO-worthy.

Actually, that doesn't happen in a single server  environment because we share the cache.  I will note the multi-server possibility though.